libr/core/cmd_search.c

```diff
@@ -1549,6 +1550,134 @@ static void do_esil_search(RCore *core, struct search_parameters *param, const c
 	r_cons_clear_line (1);
 }
 
+static void do_syscall_search(RCore *core, struct search_parameters *param) {
+	RSearch *search = core->search;
+	ut64 at;
+	ut64 curpc;
+	ut8 *buf;
+	ut8 *arr;
+	RAnalOp aop;
+	int i,j,inslen, ret, bsize = R_MIN (64, core->blocksize);
+	int kwidx = core->search->n_kws;
+	int count = 0;	
+	int nb_opcodes = 8;
+	ut64 old_offset = core->offset;
+	int nbytes = 0;
+	RAnalEsil *esil = core->anal->esil;
+	RIOMap* map;
+	RListIter *iter; 
+	int stacksize = r_config_get_i (core->config, "esil.stack.depth");
+	int bits   = r_config_get_i (core->config, "asm.bits");
+	int iotrap = r_config_get_i (core->config, "esil.iotrap");
+	
+	if (!(esil = r_anal_esil_new (stacksize, iotrap))) {
+		return;
+	}	
+	buf = malloc (bsize);	
+	arr = malloc (bsize);
+	if (!buf || !arr) {
+		eprintf ("Cannot allocate %d bytes\n", bsize);
+		return;
+	}
+	r_cons_break_push (NULL, NULL);
+	r_list_foreach (param->boundaries, iter, map) {
+		ut64 from = map->itv.addr;
+		ut64 to = r_itv_end (map->itv);
+		for (i = 0, at = from; at < to; at++, i++) {
+			if (r_cons_is_breaked ()) {
+				break;
+			}
+			if (i >= (bsize - 32)) {
+				i = 0;
+			}
+			if (!i) {
+				r_core_read_at (core, at, buf, bsize);
+			}
+			ret = r_anal_op (core->anal, &aop, at, buf + i, bsize - i);
+			if (ret) {
+				bool match = false;
+				if (aop.type == R_ANAL_OP_TYPE_SWI) {
+					match = true;
+				}
+				if (match) {
+					// This for calculating no of bytes to be subtracted , to get 8 instr above syscall
+					if (r_core_prevop_addr (core, core->offset, nb_opcodes, &core->offset)) {
+						nbytes = old_offset - core->offset;
+					} else {
+						core->offset = old_offset;
+						r_core_asm_bwdis_len (core, &nbytes, &core->offset, nb_opcodes);
+					}
+					curpc = (at <= 8)? 0 : at-nbytes; // A quick fix for small binaries
+					r_core_cmdf (core, "ar PC=0x%08"PFMT64x"", curpc);
+					r_core_cmd0 (core, ".ar*");
+					// This loop is for emulating 8 instr above the syscall 
```

Comment:
8 instructions? Expect them to be alignes at any specific size? On x86 instructions are not fixed size. U shoukd keep track of the size of each instruction instead of using bwdis and assume that is fine. Forward disasm is more reliable. Just keep a list of nth ints to save the size and drop the first item of the list on each push

Reply:
fine i will create a list then !!!
But how many instructions do u think i have to emulate before the syscall ?  coz most of the time rax value will be set within 8 instr in vicinity to syscall